src/adapt.rs

```diff
@@ -0,0 +1,481 @@
+    TypeError(TypeError),
+
+    DuplicateAdapter(Id<AdapterId>),
+    NoSuchAdapter(Id<AdapterId>),
+    InvalidValue
+}
+
+/// A witness that we are currently watching for a value.
+/// Watching stops when the guard is dropped.
+pub trait WatchGuard {
+}
+
+/// Phantom witness, used solely to define Id<AdapterId>.
+#[derive(Clone, Debug)]
+pub struct AdapterId;
+
+/// API that adapters must implement
+pub trait Adapter: Send {
+    /// An id unique to this adapter. This id must persist between
+    /// reboots/reconnections.
+    fn id(&self) -> Id<AdapterId>;
+
+    /// The name of the adapter.
+    fn name(&self) -> &str;
+    fn vendor(&self) -> &str;
+    fn version(&self) -> &[u32;4];
+    // ... more metadata
+
+    /// Request a value from a channel. The FoxBox (not the adapter)
+    /// is in charge of keeping track of the age of values.
+    fn get_values(&self, set: Vec<Id<Getter>>) -> Vec<Result<Option<Value>, Error>>;
+
+    /// Request that a value be sent to a channel.
+    fn put_values(&self, values: Vec<(Id<Setter>, Value)>) -> Result<(), Error>;
+
+    fn register_watch(&self, id: Id<Getter>, threshold: Option<Value>, cb: Box<Fn(Value) + Send>) -> Result<Box<WatchGuard>, Error>;
+}
+
+struct AdapterData {
+    /// The implementation of the adapter.
+    adapter: Box<Adapter>,
+
+    /// The services for this adapter.
+    services: HashSet<Id<ServiceId>>,
+}
+impl AdapterData {
+    fn new<T>(adapter: T) -> Self
+        where T: Adapter + 'static {
+        AdapterData {
+            adapter: Box::new(adapter),
+            services: HashSet::new(),
+        }
+    }
+}
+
+pub struct AdapterControlState {
+    /// The list of all adapter implementations.
+    adapters: HashMap<Id<AdapterId>, AdapterData>,
+
+    /// The list of all services.
+    services: HashMap<Id<ServiceId>, Service>,
+    getters: HashMap<Id<Getter>, Id<ServiceId>>,
+    setters: HashMap<Id<Setter>, Id<ServiceId>>,
+}
+pub struct AdapterControl {
+    // FIXME: Consider making the API async and moving AdapterControlBackend OMT.
+    state: Arc<Mutex<AdapterControlState>>
+}
+impl AdapterControl {
+    /// Create an empty AdapterControl.
+    /// This function does not attempt to load any state from the disk.
+    pub fn new() -> Self {
+        AdapterControl {
+            state: Arc::new(Mutex::new(AdapterControlState {
+                adapters: HashMap::new(),
+                services: HashMap::new(),
+                getters: HashMap::new(),
+                setters: HashMap::new(),
+            }))
+        }
+    }
+
+    /// Auxiliary function to remove an adapter, once the mutex has been acquired.
+    /// Clients should rather use Adapt::remove_adapter.
+    fn aux_remove_adapter(id: Id<AdapterId>, state: &mut AdapterControlState) -> Result<(), Error> {
+        let adapters = &mut state.adapters;
+        let services = &mut state.services;
+        let getters = &mut state.getters;
+        let setters = &mut state.setters;
```

Comment:
It's a bit annoying that we have a flat structure with no direct composition of adapter->1..N service->(1..N getter, 1..N setter). Would that not simplify a lot these functions?